Topics/Tech_Stacks/fine-tuning-ml/fine-tuning.md

```diff
@@ -0,0 +1,84 @@
+
+Fine-tuning a machine learning model can be described as the process of using a machine learning model that has already been trained on some task and further training that model to a similar task. 
+For instance, one could fine-tune a model that has been pretrained on general image recognition tasks and further train it to recognize different breeds of dogs.
+This works because a pre trained model that may have been trained to a high accuracy on a large dataset can be thought of a form of feature extraction; the weights of the pre-trained model has likely been trained to recognize various features of the dataset, and using that pre-trained model as a starting set may yield better results than just training a model from scratch.
+
+This may yield several advantages: 
+- Fine-tuning leverages the learning other parties have trained the model to recognize, often taking advantage of learning from larger, more optimized datasets.
+- Fine-tuning may take significantly less time to train, as the process of fine-tuning often involves 'freezing' most of the weights of the model.
+- Fine-tuning a model may be resistant to overfitting to training data, as most of the weights of the model have been 'frozen' at training time.
+- Fine-tuning enables models trained on general, large-scoped tasks to be adapted to smaller-scale or similar tasks without starting from scratch.
+
+However, this could come with some downsides.
+- Fine-tuning may cause the model to forget previously learned features or tasks if the new task is substantially different from the original task the model was trained on.
+- Fine-tuning requires a sufficient amount of data from the target domain to avoid overfitting. In cases of extremely limited data, fine-tuning may not be effective.
+- If the pre-trained model is significantly different from the target domain, fine-tuning may not yield significant improvements, and in some cases, it may even degrade performance.
+- Because the process of fine tuning often involves 'freezing' most of the weights of the model during the training process, fine-tuning such a model may limit how flexible such a model is during training.
+
+If a user believes that model fine-tuning is the right choice for their particular task, they can get started on finding a dataset and pretrained model pair that fits their task.
+
+## Finding a Pretrained Model
+
+Generally, when finding a pretrained model, a model that 
+- Has been trained on a similar dataset/task
+- Has achieved high accuracy
+- Is accessible
+is desired.
+
+A good place to look for pretrained models is [Hugging Face](https://huggingface.co/models).
+
+For example, using the Hugging face Transformers library, it is possible to import a BERT model from Hugging face using the following pieces of code:
+
+![Importing the transformers library](./transformerslib.png)
```

Comment:
Format these images better. It was hard for me to tell these are two separate images